lib/ansible/modules/windows/win_get_url.ps1

```diff
@@ -19,120 +20,194 @@
+
+        $webLastMod = $webResponse.GetResponseHeader("Last-Modified")
+    } Catch [System.Net.WebException] {
+        $result.status_code = $_.Exception.Response.StatusCode
+        Fail-Json -obj $result -message "Error requesting '$url'. $($_.Exception.Message)"
+    } Catch {
+        Fail-Json -obj $result -message "Error when requesting 'Last-Modified' date from '$url'. $($_.Exception.Message)"
+    }
+    $result.status_code = $webResponse.StatusCode
+    $result.msg = $webResponse.StatusDescription
+    $webResponse.Close()
+
+    if ($webLastMod -and ((Get-Date -Date $webLastMod) -lt $fileLastMod)) {
+        return $false
+    } else {
+        return $true
+    }
+}
+
+
+Function Download-File($result, $url, $dest, $headers, $credentials, $timeout, $use_proxy, $proxy, $whatif) {
+
+    # Check $dest parent folder exists before attempting download, which avoids unhelpful generic error message.
+    $dest_parent = Split-Path -LiteralPath $dest
+    if (-not (Test-Path -LiteralPath $dest_parent -PathType Container)) {
+        Fail-Json -obj $result -message "The path '$dest_parent' does not exist for destination '$dest', or is not visible to the current user.  Ensure download destination folder exists (perhaps using win_file state=directory) before win_get_url runs."
+    }
+
+    $webClient = New-Object System.Net.WebClient
+
+    foreach ($header in $headers.GetEnumerator()) {
+        $webClient.Headers.Add($header.Name, $header.Value)
+    }
+
+    if ($timeout) {
+        $webClient.Timeout = $timeout * 1000
+    }
+
+    if (-not $use_proxy) {
+        # Ignore the system proxy settings
+        $webClient.Proxy = $null
+    } elseif ($proxy) {
+        $webClient.Proxy = $proxy
+    }
+
+    if ($credentials) {
+        $webClient.Credentials = $credentials
+    }
+
+    Try {
+        if (-not $whatif) {
+            $webClient.DownloadFile($url, $dest)
+        }
+        $result.changed = $true
+    } Catch [System.Net.WebException] {
+        $result.status_code = $_.Exception.Response.StatusCode
+        Fail-Json -obj $result -message "Error downloading '$url' to '$dest'. $($_.Exception.Message)"
+    } Catch {
+        Fail-Json -obj $result -message "Unknown error downloading '$url' to '$dest'. $($_.Exception.Message)"
+    }
+    # FIXME: Reimplement DownloadFile() so we get the real information
+    $result.status_code = 200
+    $result.msg = 'OK'
+    $result.dest = $dest
+}
+
+
 $params = Parse-Args $args -supports_check_mode $true
 $check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
 
 $url = Get-AnsibleParam -obj $params -name "url" -type "str" -failifempty $true
 $dest = Get-AnsibleParam -obj $params -name "dest" -type "path" -failifempty $true
+$timeout = Get-AnsibleParam -obj $params -name "timeout" -type "int" -default 10
+$headers = Get-AnsibleParam -obj $params -name "headers" -type "dict"
```

Comment:
You will need to set the default to `@{}` so when call `GetEnumerator()` in the methods above it won't fail if a user doesn't pass anything in.

Reply:
I assumed that the default would adapt to the type in Ansible, but we don't do that anywhere.
I learned something new today :-)

Reply:
`dict` isn't a real type right now, even if it was I'm not sure if it should. There may be a reason where we need to know if it is null vs when it is an empty dict.